internal/inventory/tagprocessor.go

```diff
@@ -0,0 +1,364 @@
+	statsAtReadLoc := tag.getStats(readLocation.String())
+
+	if rssi, hasRSSI := rt.ExtractRSSI(); hasRSSI {
+		statsAtReadLoc.updateRSSI(rssi)
+	}
+
+	if hasTimestamp {
+		statsAtReadLoc.updateLastRead(lastRead)
+	}
+
+	if prevLoc.IsEmpty() || tag.Location.Equals(readLocation) {
+		tag.Location = readLocation
+		return
+	}
+
+	statsAtPrevLoc := tag.getStats(tag.Location.String())
+	if statsAtPrevLoc.rssiCount() == 0 {
+		// Its stats have been cleared; update location.
+		tag.Location = readLocation
+		return
+	}
+
+	// if the incoming read's location has at least 2 data points, lets see if the tag should move
+	if statsAtReadLoc.rssiCount() >= 2 {
+		if tp.config.debugLogEnabled {
+			logReadTiming(tp, info, statsAtPrevLoc, tag)
+		}
+
+		locationMean := statsAtPrevLoc.rssiDbm.Mean()
+		incomingMean := statsAtReadLoc.rssiDbm.Mean()
+
+		offset := tp.config.profile.computeOffset(info.referenceTimestamp, statsAtPrevLoc.LastRead)
+		if tp.config.debugLogEnabled {
+			logTagStats(tp, tag, readLocation.String(), incomingMean, locationMean, offset)
+		}
+
+		// Update the location if the mean RSSI at the new location
+		// is greater than the adjusted mean RSSI of the existing location.
+		// Note: This will generate a moved event.
+		if incomingMean > (locationMean + offset) {
+			tag.Location = readLocation
+		}
+	}
+
+	return
+}
+
+func logTagStats(tp *TagProcessor, tag *Tag, readLocation string, incomingMean float64, existingMean float64, offset float64) {
+	tp.lc.Debug("tag stats",
+		"epc", tag.EPC,
+		"readLoc", readLocation,
+		"prevLoc", tag.Location,
+		"incomingAvg", fmt.Sprintf("%.2f", incomingMean),
+		"existingAvg", fmt.Sprintf("%.2f", existingMean),
+		"offset", fmt.Sprintf("%.2f", offset),
+		"existingAdjusted", fmt.Sprintf("%.2f", existingMean+offset),
+		// if stayFactor is positive, tag will stay, if negative, generates a moved event
+		"stayFactor", fmt.Sprintf("%.2f", (existingMean+offset)-incomingMean))
+}
+
+func logReadTiming(tp *TagProcessor, info ReportInfo, locationStats *TagStats, tag *Tag) {
+	now := UnixMilliNow()
+	tp.lc.Debug("read timing",
+		"now", now,
+		"referenceTimestamp", info.referenceTimestamp,
+		"nowMinusRef", fmt.Sprintf("%v", time.Duration(now-info.referenceTimestamp)*time.Millisecond),
+		"locationLastRead", locationStats.LastRead,
+		"lastRead", tag.LastRead,
+		"diff", fmt.Sprintf("%v", time.Duration(tag.LastRead-locationStats.LastRead)*time.Millisecond))
+}
+
+// AgeOut is a cleanup method that will remove tag information from our in-memory
+// structures if it has not been seen in a long enough time. Only applies to
+// tags which are already Departed.
+func (tp *TagProcessor) AgeOut() (int, []StaticTag) {
+	expiration := UnixMilli(time.Now().Add(time.Hour * time.Duration(-tp.config.ageOutHours)))
+
+	// developer note: Go allows us to remove from a map while iterating
+	var numRemoved int
+	for epc, tag := range tp.inventory {
+		if tag.state == Departed && tag.LastRead < expiration {
```

Comment:
This does make sense , would think would be `>` but maybe expiration is not the right term.??? Very confusing. 

Seem clear if determine duration for last read (Time.Now - tag.LastRead) and compare that to duration for ageOutHours

Reply:
Yeah the term `expiration` was probably a confusing choice. The math is done backwards from what you typically expect, probably due to a micro-optimization on how the comparisons are made.

Instead of calculating the `expiration` as a duration of time -- which requires computing `now - tag.LastRead` for each read in order to compare on the basis of duration -- I am calculating the earliest absolute timestamp that would not be considered aged-out. Anything before that is too old. This allows me to simply compute one value and then compare each tag without applying any transformation to it (subtracting from now).